### PR TITLE
Move DataSourcePicker implementation to core

### PR DESCRIPTION
**Pull request by ivanortegaalba**
**Problem**
* The DSPicker implementation lives in `@grafana/runtime`
* From dashboards, we want to refactor that component to improve discoverability and the experience
* This involves installing dependencies like `emotion` that are not needed for the purpose of `@grafana/runtime` since this package is not meant to expose components. 

**Solution**
* Move implementation into core, [as we do with `PanelRenderer` or `PluginPage`](https://github.com/grafana/grafana/pull/66408/files#diff-ec82b7824e0c1754dc1b9a0b73fa92812e888c44bf6c9039a1fccdeffe27e812R127-R130)
* Use the component as a singleton and allow a setter to replace implementation. 
* When initializing the app, replace the implementation of the component with the one existing in core

**Consequences**
* No extra packages are installed in `@grafana/runtime`
* The implementation lives in core and it can use all the components available there 
* If the component is used before the app is initialized, there is a mocked component rendering instead
* This follows the same pattern [as other components](https://github.com/grafana/grafana/pull/66408/files#diff-ec82b7824e0c1754dc1b9a0b73fa92812e888c44bf6c9039a1fccdeffe27e812R127-R130) living in core

**Notes for the reviewer**
* I declared `current` as optional since [there are use cases](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/tempo/configuration/ServiceGraphSettings.tsx#L30) where the value received is undefined, not always null. I try to understand why TS was not complaining before.

### CODE REVIEW

**Review by dprokop — APPROVED** [review]
As long as @grafana/plugins-platform-frontend is OK with this approach, I'm game.

**Review by ryantxu — APPROVED** [review]

File: packages/grafana-runtime/src/components/DataSourcePicker.tsx [hunk 1]
```diff
@@ -23,7 +9,7 @@ import { ExpressionDatasourceRef } from './../utils/DataSourceWithBackend';
  */
 export interface DataSourcePickerProps {
   onChange: (ds: DataSourceInstanceSettings) => void;
-  current: DataSourceRef | string | null; // uid
+  current?: DataSourceRef | string | null; // uid
```
Comment (ryantxu) [inline packages/grafana-runtime/src/components/DataSourcePicker.tsx:12]: 👍

File: .betterer.results [hunk 2]
```diff
@@ -749,10 +749,6 @@ exports[`better eslint`] = {
     "packages/grafana-runtime/src/analytics/types.ts:5381": [
       [0, 0, 0, "Unexpected any. Specify a different type.", "0"]
     ],
-    "packages/grafana-runtime/src/components/DataSourcePicker.tsx:5381": [
-      [0, 0, 0, "Use data-testid for E2E selectors instead of aria-label", "0"],
-      [0, 0, 0, "Use data-testid for E2E selectors instead of aria-label", "1"]
-    ],
```
Comment (nmarrs) [inline .betterer.results:752]: pushed up a commit to resolve betterer complaints 😄

**Review by nmarrs — APPROVED** [review]
LGTM, added a commit to resolve betterer complaints / updated on latest main

**Comment by ivanortegaalba** [comment]
🤔 With this change, we're breaking every unit test that renders a component that imports `DataSourcePicker` from `@grafana/runtime`. 

Should we set `DataSourcePicker` in the `setupTest.ts` or does the setter needs to be used in every test?

IMO, I'd like to hide all of this wiring, so I'd go for adding it into `setupTest.ts`. I wonder if we're breaking also tests in plugins.

File: .betterer.results (hunk 2 above)
Comment (ivanortegaalba) [reply .betterer.results:752]: @nmarrs the Select doesn't support `data-test-id`. It only works with `aria-label`. This is breaking units and E2Es. Let's do this in a separate PR to not disrupt this one. We're kind of rushing to have the DSPicker in place asap. Thank you so much to give it a try! 🙏

**Review by mckn — APPROVED** [review]
LGTM! Great work on this! Since this will be included in G10 maybe it also would be worth to consider moving this to Grafana UI?

**Comment by ivanortegaalba** [comment]
@dprokop @jackw @leventebalogh and I discussed yesterday this, and we decided:
* Leave the component in runtime and deprecate it. Moving it into core is a breaking change for environments out of the main app, for example, unit tests.
* Develop the new component in Core, so the API for the new component is not limited by the implementation of the current one.
* For now, there is only one use case out of Core, which is ML plugin. When deprecating from runtime, we can easily reproduce the DS picker combining a select with a service.
* If there are many other use cases in the future about having a DS picker for plugins, we can expose it properly from @grafana/ui or any other place. But runtime is meant to be for services and utils only.

**Pull request closed without merging**